packages/node_modules/@webex/internal-plugin-conversation/src/activity-thread-ordering.js

```diff
@@ -0,0 +1,331 @@
+    [ACTIVITY_TYPES.REACTION]: handleNewReaction,
+    [ACTIVITY_TYPES.REACTION_SELF]: handleNewReactionSelf,
+    [ACTIVITY_TYPES.EDIT]: handleNewEdit,
+    [ACTIVITY_TYPES.REPLY]: handleNewReply
+  }[key]);
+
+  return {
+    getActivityHandlerByKey,
+    getActivityByTypeAndId
+  };
+};
+
+/**
+ * encapsulates state and logic for managing oldest and newest activities
+ * @returns {object} setters and getters for activity state management
+ */
+export const bookendManager = () => {
+  // keep track of generator state, like what our current oldest & newest activities are
+  let oldestAct;
+  let newestAct;
+
+  const getOldestAct = () => oldestAct;
+  const getNewestAct = () => newestAct;
+
+  const setOldestAct = (act) => {
+    if (!oldestAct) {
+      oldestAct = act;
+    }
+    else if (isNewer(oldestAct, act)) {
+      oldestAct = act;
+    }
+  };
+
+  const setNewestAct = (act) => {
+    if (!newestAct) {
+      newestAct = act;
+    }
+    else if (isNewer(act, newestAct)) {
+      newestAct = act;
+    }
+  };
+
+  const setBookends = (activities) => {
+    const oldestActsFirst = sortActivitiesByPublishedDate(activities);
+
+    const newestInBatch = last(oldestActsFirst);
+    const oldestInBatch = oldestActsFirst[0];
+
+    setOldestAct(oldestInBatch);
+    setNewestAct(newestInBatch);
+  };
+
+  return {
+    setBookends,
+    getNewestAct,
+    getOldestAct
+  };
+};
+
+export const noMoreActivitiesManager = () => {
+  // used to determine if we should continue to fetch older activities
+  // must be set per iteration, as querying newer activities is still valid when all end of convo has been reached
+  let noMoreActs = false;
+  let noOlderActs = false;
+  let noNewerActs = false;
+
+  const getNoMoreActs = () => noMoreActs;
+
+  const checkAndSetNoOlderActs = (act) => {
+    if (!noOlderActs && getActivityType(act) === ACTIVITY_TYPES.CREATE) {
+      noOlderActs = true;
+    }
+  };
+
+  const checkAndSetNoNewerActs = (activities) => {
+    if (!activities || !activities.length) {
+      noNewerActs = true;
+    }
+  };
+
+  const checkAndSetNoMoreActs = (queryType) => {
+    if (
+      (queryType === NEWER && noNewerActs) ||
+      ((queryType === OLDER || queryType === INITIAL) && noOlderActs)
+    ) {
+      noMoreActs = true;
+    }
+  };
+
+  return {
+    getNoMoreActs,
+    checkAndSetNoMoreActs,
+    checkAndSetNoNewerActs,
+    checkAndSetNoOlderActs
+  };
+};
+
+export const rootActivityManager = () => {
+  const rootActivityHash = new Map();
+  const handleNewRoot = (rootAct) => {
+    setValue(rootActivityHash, rootAct.id, rootAct);
+  };
+
+  const getRootActivityHash = () => rootActivityHash;
+
+  return {
+    handleNewRoot,
+    getRootActivityHash
+  };
+};
+
+export const getLoopCounterFailsafe = () => {
+  let fetchLoopCount = 0;
+
+  return () => {
+    fetchLoopCount += 1;
+    if (fetchLoopCount > fetchLoopCountMax) {
+      throw new Error('max fetches reached');
```

Comment:
Does throwing here exist the generator gracefully? Changing it to `done` or will it throw an error in the consuming code?

Reply:
It will bubble up to caller, where they can catch and respond as they like